Add fhir-connector-example.md (#1258)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/fhir-connector/fhir-connector-example.md

# FHIR Connector Example

> Given below is a sample scenario that demonstrates how to work with resources in [FHIR](https://www.hl7.org/fhir/overview.html)(Fast Healthcare Interoperability Resources). FHIR specification is a standard for exchanging health care information electronically.The WSO2 EI FHIR connector uses the FHIR RESTFul API to interact with FHIR.

Reviewer: ```suggestion
Given below is a sample scenario that demonstrates how to work with resources in [FHIR](https://www.hl7.org/fhir/overview.html)(Fast Healthcare Interoperability Resources). FHIR specification is a standard for exchanging health care information electronically. The WSO2 EI FHIR connector uses the FHIR REST API to interact with FHIR.
```

Author: I updated this paragraph.Please check and verify.

Author: Done

> 

Reviewer 2: Shall we put a link to the API or the API doc? 
What version of API we support? 
What the connector is for : "The connector can be used to invoke FHIR operations within WSO2 EI mediation logic"

Author: Added API docs and description to the doc.

## What you'll build

> Given below is a sample API that illustrates how you can connect to a FHIR with the `init` operation and then use the `create`,`readResource`, `readSpecificResourceById`,`delete` and `update` operation to send messages to the FHIR server.

Reviewer: ```suggestion
Given below is a sample API that illustrates how you can connect to a FHIR with the `init` operation and then use the `create`,`readResource`, `readSpecificResourceById`,`delete`, and `update` operation to send messages to the FHIR server.
```

* `<fhir.init>` : It will initialize the FHIR operations.
* `<fhir.create>` : To creates a new resource in a server-assigned location, use `fhir.create` operation.
* `<fhir.readResource>` : To accesses the current contents of a resource, use `fhir.readResource` operation.
* `<fhir.readSpecificResourceById>` : To accesses the specific current contents of a resource, use `fhir.readSpecificResourceById` operation.
* `<fhir.update>` : To creates a new current version for an existing resource or creates an initial version if no resource already exists for the given id, use `fhir.update` operation.

> * `<fhir.update>` : To removes an existing resource, use `fhir.delete`operation.

Reviewer: All these should be in the reference documentation so no need to mention here.

Author: Removed this part and just mentioned only the content.

> 

Reviewer 2: These are implementation detail. Shall we talk about it like below in a `scenario mindset`. 

Given below is a sample API that illustrates how you can connect to a FHIR server and invoke operations. It exposes FHIR functionalities as a restful service. Users can invoke the API with HTTP/HTTPs with required information. 

API has the context `/resources `. It has four resources  

1. `/create` :  creates a new patient at FHIR server
2. `/read` :  retrieve information about the patient from FHIR server 
3. `/readSpecificResourceById`: <map this to the scenario is it read patient by Id? Which ID?> 
4. `/delete `: remove added patient information from FHIR server.

Author: Updated.

To know the further information about these operations please refer this link.

> > **Note**:If no id element is provided, or the value is wrong, the server SHALL respond with a HTTP 400 error code, and SHOULD provide an operation outcome identifying the issue.

Reviewer: ```suggestion
> **Note**: If no ID element is provided, or the value is wrong, the server responds with a HTTP 400 error code and provides an operation outcome identifying the issue.
```

Author: Updated.

The following diagram illustrates all the required functionality of the FHIR API service that you are going to build.

> <img src="/assets/img/connectors/FHIRConnector.png" title="FHIR Connector" width="800" alt="FHIR Connector"/>

Reviewer: Add some content to describe this diagram and what is happening in this scenario.

Author: Added.

## Configure the connector in WSO2 Integration Studio

Follow these steps to set up the ESB Solution Project and the Connector Exporter Project.

{!references/connectors/importing-connector-to-integration-studio.md!}

> 1. Right click on the created ESB Solution Project and select, -> New -> Rest API to create the REST API.

Reviewer: ```suggestion
1. Right click on the created ESB Solution Project and select **New** -> **Rest API** to create the REST API.
```

Author: Updated.

> 2. Specify the API name as SendisoTestAPI and API context as /resources. You can go to the source view of the XML configuration file of the API and copy the following configuration (source view).

Reviewer: ```suggestion
2. Specify the API name as `SendisoTestAPI` and API context as `/resources`. You can go to the source view of the XML configuration file of the API and copy the following configuration (source view).
```

Author: Updated.

               </fhir.delete>
               <log level="full" separator=","/>
               <send/>
           </inSequence>
           <outSequence/>
           <faultSequence/>
       </resource>
       <resource methods="POST" url-mapping="/update">
           <inSequence>
               <property expression="json-eval($.base)" name="base" scope="default" type="STRING"/>
               <property expression="json-eval($.resourceType)" name="type" scope="default" type="STRING"/>
               <property expression="json-eval($.format)" name="format" scope="default" type="STRING"/>
               <property expression="json-eval($.idToUpdate)" name="idToDelete" scope="default" type="STRING"/>
               <fhir.init>
                   <base>{$ctx:base}</base>
               </fhir.init>
               <fhir.update>
                   <type>{$ctx:type}</type>
                   <idToUpdate>{$ctx:idToUpdate}</idToUpdate>
                   <format>{$ctx:format}</format>
               </fhir.update>
               <log level="full" separator=","/>
               <send/>
           </inSequence>
           <outSequence/>
           <faultSequence/>
       </resource>
   </api>

>     ```

Reviewer 2: Here, point to the reference doc 
To learn about supported operations and their parameters, please refer to `FHIR connector reference`.

Author: Added.

> 3. Now we can export the imported connector and the API into a single CAR application. CAR application is the one we are going to deploy to server runtime.

Reviewer: ```suggestion
3. Now we can export the imported connector and the API into a single CAR application. The CAR application is what we are going to deploy during server runtime.
```

Author: Updated.

{! /references/connectors/exporting-artifacts.md !}

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}
    
## Testing

Invoke the API as shown below using the curl command. Curl Application can be downloaded from [here](https://curl.haxx.se/download.html).

> 1. Create resource operation.

Reviewer 2: Let's say wha we are going to do here.  Example : 
1. Add patient information

Author: Added titles as mentioned above.

   
   **Sample Request** 
      
   ```
   curl -v POST -d 
   '{

>       "base": "http://hapi.fhir.org/baseR4",

Reviewer 2: Should user provide "base"? We can hardcode it. 
What are the allowed formats? Can we inject it using `Content-Type` (using switch mediator we can set the property).

Author: If user need to use another FHIR test server or API for their scenarios they should provide base URL as this parameter.

Allowed formats of the request should be the Mime Type.

Reviewer: We should add a note on this.

Author: - Added switch mediator and update the sample API source. 
- While hardcoding the resourceType parameter and invoking the API requests, following issue occurred.
https://github.com/wso2/micro-integrator/issues/1350

      "resourceType": "Patient",
      "format": "json",
      "name": [{"family": "Jhone","given": ["Winney","Rodrigo"]}]
    }' "http://localhost:8290/resources/create" -H "Content-Type:application/json" 
       
   ```
   **Expected Response**: 
      
   ```
   <jsonObject>
       <resourceType>Patient</resourceType>
       <id>698021</id>
        <meta>
          <versionId>1</versionId>
          <lastUpdated>2020-03-24T07:57:14.506+00:00</lastUpdated>
        </meta>
         <text>
           <status>generated</status>

>            <div>&lt;div xmlns="http://www.w3.org/1999/xhtml"&gt;&lt;div class="hapiHeaderText"&gt;Winney Rodrigo &lt;b&gt;JHONE &lt;/b&gt;&lt;/div&gt;&lt;table class="hapiPropertyTable"&gt;&lt;tbody/&gt;&lt;/table&gt;&lt;/div&gt;</div>

Reviewer 2: What information is included here?

Author: While invoking each operations it already getting same content from the HAPI test server.
Seems it is a part of response body of base URL.Please refer [1].

[1] http://hapi.fhir.org/baseR4

Reviewer: Let's include this as a note as well.

         </text>
         <name>
            <family>Jhone</family>
            <given>Winney</given>
            <given>Rodrigo</given>
        </name>
   </jsonObject>
   ```  
           

> 2. readResource operation.

Reviewer 2: Same here, we'll add what we are going to do

Author: Changed.

# FHIR Connector Example

Fast Healthcare Interoperability Resources (FHIR) is an interoperability standard for electronic exchange of healthcare information. The WSO2 EI FHIR connector can be used to invoke FHIR operations within WSO2 EI mediation logic. 

> In this example the connector uses the FHIR RESTFul Api  to interact with FHIR. This connector uses the [HAPI FHIR APIs](https://hapifhir.io) to connect to a with Test Server. Which is a Open Source, licensed under the Apache Software License 2.0 (Java based implementation of the FHIR specification). 

Reviewer: ```suggestion
In this example the connector uses the FHIR REST API  to interact with FHIR. This connector uses the [HAPI FHIR APIs](https://hapifhir.io) to connect with a Test Server, which is an open source server licensed under the Apache Software License 2.0 (Java-based implementation of the FHIR specification). 
```

Author: Done
